nomad/state/paginator/paginator.go

```diff
@@ -21,136 +20,132 @@ type Iterator interface {
 }
 
 // Paginator wraps an iterator and returns only the expected number of pages.
-type Paginator struct {
+type Paginator[T, TStub any] struct {
 	iter           Iterator
-	tokenizer      Tokenizer
-	filters        []Filter
+	tokenizer      Tokenizer[T]
+	bexpr          *bexpr.Evaluator
+	filter         FilterFunc[T]
+	stubFn         func(T) (TStub, error)
 	perPage        int32
 	itemCount      int32
-	seekingToken   string
-	seekingUint    uint64
 	nextToken      string
 	reverse        bool
 	nextTokenFound bool
 	pageErr        error
-
-	// appendFunc is the function the caller should use to append raw
-	// entries to the results set. The object is guaranteed to be
-	// non-nil.
-	appendFunc func(interface{}) error
 }
 
 // NewPaginator returns a new Paginator. Any error creating the paginator is
 // due to bad user filter input, RPC functions should therefore return a 400
 // error code along with an appropriate message.
-func NewPaginator(iter Iterator, tokenizer Tokenizer, filters []Filter,
-	opts structs.QueryOptions, appendFunc func(interface{}) error) (*Paginator, error) {
-
-	var evaluator *bexpr.Evaluator
-	var err error
+func NewPaginator[T, TStub any](
+	iter Iterator,
+	opts structs.QueryOptions,
+	filter FilterFunc[T],
+	tokenizer Tokenizer[T],
+	stubFn func(T) (TStub, error),
+) (*Paginator[T, TStub], error) {
+
+	p := &Paginator[T, TStub]{
+		iter:           iter,
+		tokenizer:      tokenizer,
+		filter:         filter,
+		stubFn:         stubFn,
+		perPage:        opts.PerPage,
+		reverse:        opts.Reverse,
+		nextTokenFound: opts.NextToken == "",
+	}
 
 	if opts.Filter != "" {
-		evaluator, err = bexpr.CreateEvaluator(opts.Filter)
+		evaluator, err := bexpr.CreateEvaluator(opts.Filter)
 		if err != nil {
 			return nil, fmt.Errorf("failed to read filter expression: %v", err)
 		}
-		filters = append(filters, evaluator)
+		p.bexpr = evaluator
 	}
 
-	// attempt to convert token to uint for iterators ordered numerically.
-	// it's safe to ignore the error here because the `next` method ignores
-	// this field for string tokens and 0 is valid for an unset numeric token.
-	seekingUint, _ := strconv.ParseUint(opts.NextToken, 10, 64)
-
-	return &Paginator{
-		iter:           iter,
-		tokenizer:      tokenizer,
-		filters:        filters,
-		perPage:        opts.PerPage,
-		seekingToken:   opts.NextToken,
-		seekingUint:    seekingUint,
-		reverse:        opts.Reverse,
-		nextTokenFound: opts.NextToken == "",
-		appendFunc:     appendFunc,
-	}, nil
+	return p, nil
 }
 
+// func (p *Paginator[T, TStub]) WithFilter(fn FilterFunc[T]) *Paginator[T, TStub] {
+// 	p.filter = fn
+// 	return p
+// }
```

Comment:
🤔

Reply:
Whoops, forgot to clean out my original idea! :blush: